service/history/replication/eventhandler/local_events_handler.go

```diff
@@ -194,7 +220,14 @@ func (h *localEventsHandlerImpl) importEvents(
 
 		blobSize++
 		blobs = append(blobs, batch.RawEventBatch)
-		if blobSize >= historyImportBlobSize || len(blobs) >= historyImportPageSize {
+		events, err := h.eventSerializer.DeserializeEvents(batch.RawEventBatch)
+		if err != nil {
+			return err
+		}
+
+		if blobSize >= historyImportBlobSize ||
+			len(blobs) >= historyImportPageSize ||
+			h.isLastEventAtHistoryBoundary(events[len(events)-1], versionHistory) { // Import API only take events that has same version
```

Comment:
where do we check the events must have same version?

Reply:
It's here: https://github.com/temporalio/temporal/blob/00fab090ff656dcff2a9e669e7835cdd6e893dbf/service/history/ndc/replication_task.go#L518